modules/contracts/apps/subscription/src/handlers/execute.rs

```diff
@@ -0,0 +1,316 @@
+                Err(SubscriptionError::EmissionsAlreadyClaimed {})
+                | Err(SubscriptionError::SubscriberEmissionsNotEnabled {}) => None,
+                Err(error) => {
+                    return Err(error);
+                }
+            };
+
+            subscription_state.active_subs -= 1;
+            SUBSCRIBERS.remove(deps.storage, &addr);
+            EXPIRED_SUBSCRIBERS.save(deps.storage, &addr, &subscriber)?;
+            canceled_subs.push(addr.into_string());
+
+            if let Some(msg) = maybe_claim_msg {
+                claim_msgs.push(msg)
+            }
+        }
+    }
+
+    // Error if no one unsubbed
+    if canceled_subs.is_empty() {
+        return Err(SubscriptionError::NoOneUnsubbed {});
+    }
+
+    // update subscription count
+    SUBSCRIPTION_STATE.save(deps.storage, &subscription_state)?;
+
+    let mut response = app.tag_response(Response::new().add_messages(claim_msgs), "unsubscribe");
+
+    if let Some(hook) = subscription_config.unsubscription_hook_addr {
+        let msg = UnsubscribedHookMsg {
+            unsubscribed: canceled_subs,
+        }
+        .into_cosmos_msg(hook)?;
+        response = response.add_message(msg);
+    }
+
+    Ok(response)
+}
+
+// Claim emissions
+pub fn claim_emissions_msg(
+    app: &SubscriptionApp,
+    deps: Deps,
+    env: &Env,
+    subscriber: &mut Subscriber,
+    subscriber_addr: &Addr,
+    subscription_per_week_emissions: EmissionType<Addr>,
+    subscription_state: &SubscriptionState,
+) -> SubscriptionResult<Option<ExecutorMsg>> {
+    if subscriber.last_emission_claim_timestamp >= env.block.time {
+        return Err(SubscriptionError::EmissionsAlreadyClaimed {});
+    }
+
+    let duration = env
+        .block
+        .time
+        .minus_seconds(subscriber.last_emission_claim_timestamp.seconds());
+    let weeks_passed = duration.seconds() / WEEK_IN_SECONDS;
+
+    let asset = match subscription_per_week_emissions {
+        crate::state::EmissionType::None => {
+            return Err(SubscriptionError::SubscriberEmissionsNotEnabled {});
+        }
+        crate::state::EmissionType::WeekShared(shared_emissions, token) => {
+            // active_sub can't be 0 as we already loaded from storage
+            let amount = (shared_emissions * Uint128::from(weeks_passed))
+                / (Uint128::from(subscription_state.active_subs));
+            Asset::new(token, amount)
+        }
+        crate::state::EmissionType::WeekPerUser(per_user_emissions, token) => {
+            let amount = per_user_emissions * Uint128::from(weeks_passed);
+            Asset::new(token, amount)
+        }
+    };
+
+    if !asset.amount.is_zero() {
+        // Update only if there was claim
+        subscriber.last_emission_claim_timestamp = env.block.time;
+
+        let send_msg = app.bank(deps).transfer(vec![asset], subscriber_addr)?;
+        Ok(Some(app.executor(deps).execute(vec![send_msg])?))
```

Comment:
this should return AccountAction so you can combine multiple of them into one CosmosMsg. Saves a lot of gas!